classes/mail/mailables/DecisionBackToInternalFromExternalReviewNotifyReviewer.inc.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+/**
+ * @file classes/mail/mailables/DecisionBackToInternalFromExternalReviewNotifyReviewer.inc.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class DecisionBackToInternalFromExternalReviewNotifyReviewer
+ *
+ * @brief Email sent to the author(s) when the following decision is made:
+ *   SUBMISSION_EDITOR_DECISION_BACK_TO_INTERNAL_REVIEW_FROM_EXTERNAL_REVIEW
+ */
+
+namespace PKP\mail\mailables;
+
+use APP\decision\Decision;
+use APP\submission\Submission;
+use PKP\context\Context;
+
+class DecisionBackToInternalFromExternalReviewNotifyReviewer extends DecisionNotifyReviewer
```

Comment:
Done. I created a new Mailable class `DecisionReviewerUnassignedNotifyReviewer` and removed other mailables.

Reply:
We've still got `PKP\mail\mailables\ReviewerUnassigned` and I think it makes sense to re-use that mailable, rather than creating a separate one for the decision.

Reply:
@NateWr we can use the mailable `PKP\mail\mailables\ReviewerUnassign` but in the `PKP\decision\types\traits\NotifyReviewers` trait , it's defined as 

```php
protected function sendReviewersEmail(DecisionNotifyReviewer $mailable, EmailData $email, User $editor, Submission $submission)
```

wondering is there any specific reason to mark the `$mailable` as `DecisionNotifyReviewer` instead of top level class `PKP\mail\Mailable` ? if not then I can update the type to `PKP\mail\Mailable` and we can reuse the already existing one `ReviewerUnassign` ? otherwise we need to have a different similar implementation to reuse it .

Reply:
It looks like `DecisionNotifyReviewer` has an extra method, `getDecision()`, that is required in `sendReviewersEmail`. This is used to update the review assignment to indicate that the reviewer has been thanked for their review.

Looking at `sendReviewersEmail()` and `validateNotifyReviewersAction()`, I wonder if the `notifyReviewers` trait is maybe not appropriate for the decisions to cancel review rounds. In the first decisions, reviewers are being notified of a result from a _completed_ review. In the latter decisions, we want to cancel _uncompleted_ review assignments.

I think `DecisionBackToInternalFromExternalReviewNotifyReviewer` and similar decisions probably shouldn't use the `notifyReviewers` trait.

Reply:
I noticed that the docblock for the `notifyReviewers` trait is unhelpful. Maybe change it to this:

> @brief Helper functions for decisions that notify reviewers of a decision taken as a result of their completed review.

Reply:
> I noticed that `validateNotifyReviewersAction()` calls `getCompletedReviewerIds()`

@NateWr that is correct but that was before my last commit . it has been refactored to use a more generalised method now as defined in the `NotifyReviewer` trait

```php
/** @copydoc WithReviewAssignments::getReviewerIds() */
abstract protected function getReviewerIds(int $submissionId, int $reviewRoundId, int $reviewAssignmentState): array;
```

> Maybe if the trait had a `getValidReviewers` function, each decision would be forced to implement that?

I think the above implementation does it already . 

> The problem with this is that `ReviewerUnassigned` is used in places where no decision exists. An editor can unassign a reviewer without recording a decision.

you are right and I am sorry I did not notice that difference . In that sense not sure if we can use it as this `ReviewerUnassigned` is designed to handle a simple unassignment but in our case , the unassignment is associated with a decision. 

I an thinking that in the `NotifyReviewer` trait , the `validateNotifyReviewersAction` is ok and working fine . but the problem is the `sendReviewersEmail` from where mails are send to reviewers as an acknowledgement result of their review work . what if we can just force each decision to implement this method as per their need and defined it as an `abstract method` in the trait ? 

Another approach I am thinking that modifing the `lib/pkp` `DecisionNotifyReviewer` so that it take email `options(name, description, emailTemplateKey)` as the params in `constructor` and in the `getDecisionDescription` let it determined the `unassigned` case dynamically . 

in the `getDecisionDescription` method,  `DECISION::..` defined there shared between the both `OJS` and `OMP` but for the unassigned case we have few different ones here such as 

`OJS` reviewer unassigned Decision types
1. BACK_TO_SUBMISSION_FROM_EXTERNAL_REVIEW
2. BACK_TO_PREVIOUS_EXTERNAL_REVIEW_ROUND

`OMP` reviewer unassigned Decision types
1. BACK_TO_PREVIOUS_EXTERNAL_REVIEW_ROUND
2. BACK_TO_PREVIOUS_INTERNAL_REVIEW_ROUND
3. BACK_TO_SUBMISSION_FROM_EXTERNAL_REVIEW
4. BACK_TO_SUBMISSION_FROM_INTERNAL_REVIEW
5. BACK_TO_INTERNAL_REVIEW_FROM_EXTERNAL_REVIEW

so the modifications will be like in the `DecisionNotifyReviewer` mailable class such as 

```php
public function __construct(Context $context, 
                            Submission $submission, 
                            Decision $decision,
                            ?string $name = 'mailable.decision.notifyReviewer.name',
                            ?string $description = 'mailable.decision.notifyReviewer.description',
                            ?string $emailTemplateKey = 'REVIEW_ACK')
 {
    $this->decision = $decision;
    parent::__construct([$context, $submission, $decision]);

    self::$name = $name;
    self::$description = $description;
    self::$emailTemplateKey = $emailTemplateKey;
}

/**
 * Check if decision is type that cause reviewers to be unassigned
 */
protected function isDecisionUnassigned(int $decisionType, array $reviewerUnassignedTypes): ?int {

    return in_array($decisionType, $reviewerUnassignedTypes, true) ? $decisionType : null;
}

/**
 * Get a description of the decision to use as an email variable
 */
protected function getDecisionDescription(?string $locale = null): string
{
    $class = '\APP\decision\Decision';
    $reviewerUnassignedTypes = collect([
            'BACK_TO_PREVIOUS_EXTERNAL_REVIEW_ROUND',
            'BACK_TO_PREVIOUS_INTERNAL_REVIEW_ROUND',
            'BACK_TO_SUBMISSION_FROM_EXTERNAL_REVIEW',
            'BACK_TO_SUBMISSION_FROM_INTERNAL_REVIEW',
            'BACK_TO_INTERNAL_REVIEW_FROM_EXTERNAL_REVIEW',
        ])
        ->map(fn($type) => defined("{$class}::{$type}") ? constant("{$class}::{$type}") : null)
        ->filter(fn($type) => !is_null($type))
        ->toArray();

    return match($decisionType = $this->decision->getData('decision')) {
        Decision::ACCEPT => __('mailable.decision.notifyReviewer.variable.decisionDescription.accept', [], $locale),
        Decision::DECLINE => __('mailable.decision.notifyReviewer.variable.decisionDescription.decline', [], $locale),
        Decision::PENDING_REVISIONS => __('mailable.decision.notifyReviewer.variable.decisionDescription.pendingRevisions', [], $locale),
        Decision::RESUBMIT => __('mailable.decision.notifyReviewer.variable.decisionDescription.resubmit', [], $locale),
        $this->isDecisionUnassigned($decisionType, $reviewerUnassignedTypes) => __('mailable.decision.notifyReviewer.variable.decisionDescription.unassigned', [], $locale),
        default => '',
    };
}
```
with the above modification, we can use the current code base with little modification and have the `DecisionNotifyReviewer` reuse for our purpose . 

Another option is to use the `ReviewerUnassign` mailbale and passed the `reviewAssignment` in each of decision class but in that case we wont have the decision associated with it .

which one you think have better use case for use considering future enhancement and modifications ? Sorry if I am making things complicated here .

Reply:
> I an thinking that in the NotifyReviewer trait , the validateNotifyReviewersAction is ok and working fine . but the problem is the sendReviewersEmail from where mails are send to reviewers as an acknowledgement result of their review work . what if we can just force each decision to implement this method as per their need and defined it as an abstract method in the trait ?

I don't love it because in my view the point of a `trait` is to provide features. What you describe is more of a contract/interface. If we go that route, let's make it an interface or something maybe.

This is an area where maybe a union type is justified. It would let us keep `notifyReviewers` almost exactly as it is

```php
protected function sendReviewersEmail(
    DecisionNotifyReviewer|ReviewerUnassign $mailable,
    EmailData $email,
    User $editor,
    Submission $submission
)
{
    // ...

    foreach ($recipients as $recipient) {
        // ...

        if (is_a($mailable, DecisionNotifyReviewer::class)) {
            // Update the ReviewAssignment to indicate the reviewer has been acknowledged
        }
    }

    // ...
}
```

The downside is that the behaviour is implicit rather than explicit, which can sometimes make it harder to catch side effects. An alternative is to pass a flag to the function to explicitly determine behaviour.

```php
protected function sendReviewersEmail(
    DecisionNotifyReviewer|ReviewerUnassign $mailable, 
    EmailData $email, 
    User $editor, 
    Submission $submission,
    ?int $newReviewAssignmentStatus = null
)
{
    // ...

    foreach ($recipients as $recipient) {
        // ...

        if ($newReviewAssignmentStatus !== null) {
            // Update the ReviewAssignment 
        }
    }

    // ...
}
```

> Another approach I am thinking that modifing the lib/pkp DecisionNotifyReviewer so that it take email options(name, description, emailTemplateKey) as the params in constructor and in the getDecisionDescription let it determined the unassigned case dynamically .

This is not going to work because the name/description need to be able to be derived from a mailable without instantiating the class. The reason is that these variables will be used in an interface where journal managers can configure their email templates. 

So, basically, we need one `Mailable` class for every "type of event" where an email is sent. What makes something a "type of event" is not a technical question, but has to do with whether or not a user would realistically need to use a distinct template. `DecisionNotifyReviewer` may be very similar to `ReviewerUnassign` from a technical perspective. But from the user's perspective, they are very different actions and a journal wants to write different email messages to reviewers for each event.

Reply:
To follow on from my last comment about "type of event". The `DecisionNotifyReviewer` mailable is probably misleading because the "type of event" is best described like this:

> Notify a reviewer of a decision taken in response to their review.

When cancelling a review round, `DecisionNotifyReviewer` is the wrong "type of event" because the decision that is being taken is not in response to their review. That's why we need to keep distinct mailiables here.

Reply:
@NateWr the way `ReviewerUnassign` was designed, it was not possible to use it that way with the `NotifyReviewers` trait's `sendReviewersEmail` method . Also using it with `union` type and then having an extra check which make it bit too messy as we like keep the methods clean for their purpose . So I have broken it down into 3 traits `NotifyReviewersOfUnassignment`, `NotifyReviewersOfAcknowledgement` and `ToNotifyReviewers` which make it bit clean (I think so) . please review this changes also if you think the naming is not right , we can update that also .